patches/server/0726-Expose-firework-item-directly-manually-setting-fligh.patch

```diff
@@ -0,0 +1,126 @@
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+@@ -14,24 +14,26 @@ import org.bukkit.inventory.meta.FireworkMeta;
+ public class CraftFirework extends CraftProjectile implements Firework {
+ 
+     private final Random random = new Random();
+-    private final CraftItemStack item;
++    //private final CraftItemStack item; Paper - Expose firework item directly
+ 
+     public CraftFirework(CraftServer server, FireworkRocketEntity entity) {
+         super(server, entity);
+ 
+-        ItemStack item = this.getHandle().getEntityData().get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM);
+-
+-        if (item.isEmpty()) {
+-            item = new ItemStack(Items.FIREWORK_ROCKET);
+-            this.getHandle().getEntityData().set(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM, item);
+-        }
+-
+-        this.item = CraftItemStack.asCraftMirror(item);
+-
+-        // Ensure the item is a firework...
+-        if (this.item.getType() != Material.FIREWORK_ROCKET) {
+-            this.item.setType(Material.FIREWORK_ROCKET);
+-        }
++        // Paper Start - Expose firework item directly
++//        ItemStack item = this.getHandle().getEntityData().get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM);
++//
++//        if (item.isEmpty()) {
++//            item = new ItemStack(Items.FIREWORK_ROCKET);
++//            this.getHandle().getEntityData().set(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM, item);
++//        }
++//
++//        this.item = CraftItemStack.asCraftMirror(item);
++//
++//        // Ensure the item is a firework...
++//        if (this.item.getType() != Material.FIREWORK_ROCKET) {
++//            this.item.setType(Material.FIREWORK_ROCKET);
++//        }
++        // Paper End - Expose firework item directly
+     }
+ 
+     @Override
+@@ -51,17 +53,28 @@ public class CraftFirework extends CraftProjectile implements Firework {
+ 
+     @Override
+     public FireworkMeta getFireworkMeta() {
+-        return (FireworkMeta) this.item.getItemMeta();
++        // Paper Start - Expose firework item directly
++        org.bukkit.inventory.ItemStack itemStack = getItem();
++        if (itemStack.getType() != Material.FIREWORK_ROCKET) {
++            return (FireworkMeta) org.bukkit.craftbukkit.inventory.CraftItemFactory.instance().getItemMeta(Material.FIREWORK_ROCKET); // Return normal meta
++        }
++        return (FireworkMeta) itemStack.getItemMeta();
++        // Paper End - Expose firework item directly
+     }
+ 
+     @Override
+     public void setFireworkMeta(FireworkMeta meta) {
+-        this.item.setItemMeta(meta);
+-
++        // Paper Start - Expose firework item directly
++        org.bukkit.inventory.ItemStack itemStack = getItem();
++        if (itemStack.getType() != Material.FIREWORK_ROCKET) {
++            itemStack.setType(Material.FIREWORK_ROCKET);
```

Comment:
Check if the meta is applicable to the type with Bukkit.getItemFactory().isApplicable